### PR TITLE
fix(bokslut): round to öre before truncating to whole kronor

### LINKED ISSUE #2597: Whole-krona truncation of double sums drops a krona: INK2 filings and schablonavdrag
### What happened?

Three places truncate a SEK amount to whole kronor with `Math.floor` / `Math.ceil` applied straight to a sum of doubles, with no rounding to öre first. When the sum lands just below an integer, the result is a krona short. Two of the three produce a wrong declared figure.

| Site | Input drifts | Wrong output | Consequence |
|---|---|---|---|
| `ink2-engine.ts:363` | yes | **yes** | one krona short in an INK2 filing |
| `egenavgifter-calculator.ts:67` | yes | **yes** | schablonavdrag one krona low (R43 on NE-bilagan) |
| `periodiseringsfond-service.ts:152` | yes | no, in practice | the 25 % cap absorbs it, see below |

`truncateToWholeKronor` in `lib/money.ts` already guards against exactly this, and says so in its docblock. None of the three uses it.

#### 1. INK2 rutor

The chain, all in `lib/reports/ink2/ink2-engine.ts`:

1. `ink2r[code] += amount` (line 352) accumulates ledger amounts into the ruta. Exact in Postgres `numeric`, doubles once in JS.
2. `ink2r[code] = truncateToKrona(ink2r[code])` (line 363) truncates that raw sum to whole kronor.
3. `truncateToKrona` (line 83) is `value >= 0 ? Math.floor(value) : Math.ceil(value)`, with no rounding to öre first.
4. `sruAmount` writes it to the SRU file for Skatteverket.

Reproduced with three ordinary account balances:

```
1033.54 + 1203.03 + 259.43 = 2495.99999999999955   filed 2495, correct 2496
1668.37 +  210.04 + 132.59 = 2010.99999999999977   filed 2010, correct 2011
1686.00 + 1275.82 + 688.18 = 3649.99999999999955   filed 3649, correct 3650
```

#### 2. Schablonavdrag for egenavgifter

`lib/bokslut/enskild-firma/egenavgifter-calculator.ts:67`:

```ts
const netSurplus = Math.max(0, input.surplusBeforeEgenavgifter + priorSchablon - priorActual)
const schablonavdrag = Math.floor(netSurplus * rates.schablon)
```

The three-term sum drifts the same way. Reproduced against exact öre arithmetic:

```
10209.31 + 1854.55 -   43.86 = 12019.9999999999982   schablonavdrag 3004, correct 3005
 4721.11 + 1915.69 - 1804.80 =  4831.99999999999909  schablonavdrag 1207, correct 1208
 3507.45 +  466.53 -  421.98 =  3551.99999999999955  schablonavdrag  887, correct  888
```

The deduction comes out one krona low, so the sole trader is declared slightly worse off than they are. That figure is R43 on the NE-bilaga.

#### 3. Periodiseringsfond cap: same mechanism, no wrong answer today

`lib/bokslut/reserves/periodiseringsfond-service.ts:152` floors `skattemassigtResultatBeforeAvsattning`, a six-term sum built in `dispositions-proposal-builder.ts:164`, so `base` does come out a krona low:

```
taxable = 45449.9999999999927   base 45449, correct 45450
```

But the cap is `Math.floor(base * 0.25)`, and one krona of base rarely survives that. Four million random trials produced no case where the cap itself differed. Worth fixing with the others for consistency, not on its own merits.

#### Where the problem is NOT

The rate multiplication is safe, so a fix should not go looking there. Compared against exact integer arithmetic over five million whole-krona inputs, `Math.floor(kr * rate)` is correct for every rate in play: 0.25 (schablonavdrag full, periodiseringsfond), 0.10 (pensionärer) and 0.20 (passiv näringsverksamhet). The drift is entirely in the sums that feed them.

#### The same legal rule is implemented three times, and only one is guarded

Öretal bortfaller (SFF 2011:1261 22 kap. 1 §) appears as:

| Implementation | Guards against drift | Receives raw sums |
|---|---|---|
| `truncateToWholeKronor` (`lib/money.ts`) | **yes**, runs `roundOre` first | yes, and handles it |
| `truncateToKrona` (`ink2-engine.ts:83`) | no | **yes** |
| `sruAmount` (`sru/format.ts:25`) | no | no, gets pre-truncated integers |

`truncateToWholeKronor` already carries the explanation in its docblock: it rounds first "so IEEE drift just below an integer (16 073,9999999… for a true 16 074,00) cannot lose a whole krona". That is exactly what `truncateToKrona` does not do.

`sruAmount` is unguarded too but harmless today, because line 363 hands it integers. It is worth fixing anyway so the next caller cannot be surprised, but the live site is line 363.

#### Not covered by #2594

That PR corrects `roundOre` in `lib/money.ts` and touches four files, none under `lib/reports/`. `truncateToWholeKronor` benefits from it; `truncateToKrona` has its own copy of the rule and does not.

### Steps to reproduce

The engine's own accumulate-then-truncate pattern, with its own helper:

```ts
const truncateToKrona = (v: number) => v >= 0 ? Math.floor(v) : Math.ceil(v)

let ruta = 0
for (const amount of [1033.54, 1203.03, 259.43]) ruta += amount

ruta                                  // 2495.99999999999955
truncateToKrona(ruta)                 // 2495   <- what INK2 files
truncateToKrona(roundOre(ruta))       // 2496   <- correct
```

Any ruta whose contributing balances sum to a double just below an integer reproduces it. The values above were found by searching realistic two-decimal amounts, not hand-picked edge cases.

Schablonavdrag, same shape:

```ts
const netSurplus = Math.max(0, 10209.31 + 1854.55 - 43.86)  // 12019.9999999999982
Math.floor(netSurplus * 0.25)                               // 3004  <- declared
Math.floor(roundOre(netSurplus) * 0.25)                     // 3005  <- correct
```

**Not verified:** that a specific real company's books currently produce such a sum. What is verified is that the code path is unguarded, that the guard exists elsewhere for the same rule, and that ordinary balances reach the failing state. Whether any filed INK2 is already wrong needs a check against real data.

### Expected behavior

A ruta whose contributions sum to a whole krona is filed as that krona.

The narrow fix is to round to öre before truncating, exactly as `truncateToWholeKronor` does:

```ts
function truncateToKrona(value: number): number {
  const v = roundOre(value)
  return v >= 0 ? Math.floor(v) : Math.ceil(v)
}
```

The better fix is to have one implementation. `truncateToWholeKronor` is already the canonical one and is semantically identical: compared over 142 858 values from -5 000,00 to 5 000,00 they never disagree, and it additionally normalises the `-0` that `Math.floor`/`Math.ceil` leaves on small negatives. So `ink2-engine.ts` can import it and delete the local copy, and `sruAmount` can call it rather than `Math.trunc`. The two bokslut sites need the same treatment: round to öre before flooring, since `Math.floor` there is a statutory round-down rather than a truncation and cannot simply call `truncateToWholeKronor`.

Three copies of one statutory rule, with the guard on the copy that does not need it and missing from the one that does, is the shape worth removing rather than patching.

### Deployment

Hosted (app.gnubok.se)

### Browser and OS

Not browser-specific; the arithmetic is server-side.

### PR DESCRIPTION
**Pull request by valpet**
An accumulated monetary sum just below a whole krona was truncated before its floating-point drift was removed. Three ordinary account balances totaling 2496.00 therefore produced 2495 in INK2. The same mechanism understated NE R43, a periodiseringsfond cap, and the corporate-tax base.

Closes #2597.

## Changes

- INK2 and shared SRU amount formatting call `truncateToWholeKronor` directly. The local INK2 implementation and alias are removed.
- Egenavgifter normalizes the summed surplus to öre before applying the unchanged deduction rate and whole-krona floor.
- Periodiseringsfond normalizes the taxable base and already-provisioned balance before flooring, preserving the annual cap.
- Bolagsskatt normalizes the adjusted result before the existing floor to whole tens. The reproduced taxable base now remains 56110 instead of falling to 56100; both proposal legs use the same resulting tax amount.

## Why the problem occurred

Multiple implementations of whole-krona conversion had drifted apart. The shared money helper already normalized accumulated floating-point error, but the INK2 copy and raw bokslut floors did not. This PR consumes the canonical money helpers, including the `roundOre` correction already merged in #2594.

## What was removed or simplified

INK2 no longer owns a second truncation implementation or an alias for it. SRU uses the same shared helper. The three bokslut calculations keep their existing floors because those operations have a different purpose from truncation toward zero.

## Why this solution

Normalize the summed monetary input once, then apply the existing filing or deduction rule. Rounding the deduction product instead could increase a genuine fractional deduction, so the tests cover that boundary. Converting money storage and interfaces to integer öre is outside this focused defect.

These boundaries consume currency amounts. A caller supplying genuine precision finer than one öre would have that precision rounded away; no such input contract was found in the reviewed filing paths. Ordinary fractional kronor still truncate, including negative amounts.

The corporate-tax floor to whole tens is confirmed in [IL 1 kap. 7 section](https://www.riksdagen.se/sv/dokument-och-lagar/dokument/svensk-forfattningssamling/inkomstskattelag-19991229_sfs-1999-1229/). The shared helper references the whole-krona rule in [SFF 22 kap. 1 section](https://www.riksdagen.se/sv/dokument-och-lagar/dokument/svensk-forfattningssamling/skatteforfarandeforordning-20111261_sfs-2011-1261/).

## Verification

- All three INK2 repro sums, their negative counterparts, and all three egenavgifter repro sums are covered.
- Declaration tests assert both INK2R fields and the actual generated `BLANKETTER.SRU` records. Existing declaration, mapping, warning, and NE/SRU tests remain intact.
- Boundary cases cover zero, negative zero, exact integers, genuine fractional kronor, reduced deduction categories, already-provisioned headroom, and balanced corporate-tax proposal lines.
- Replacing only the five fixed source files with `origin/main` produces 17 failing tests and 43 passing tests. Restoring the fixes makes all 60 pass.
- Focused verification: 264 tests passed across 12 files.
- Independent code, security, and Swedish accounting reviews found no blocking issue.
- Zero-extension production build passed. Lint passed with zero errors; the typecheck ratchet passed with 533 existing errors and no new errors.
- Guards, skills, taxonomy, API skill, packs, registry, and extension-isolation checks passed. API skill and registry checks required temporary LF line endings to match the Linux checkout; those files have no changes in this PR.

- Full Windows unit run: 23,365 passed, 15 failed, 5 skipped, plus four test-file loading errors. An isolated `origin/main` checkout reproduced nine failures and all four loading errors across seven files (Windows paths/line endings). The remaining failures were intermittent PDF test timeouts; all 25 tests in those two files passed when rerun serially with a 30-second test timeout. No test configuration was changed.

GitHub verification on `4660356bf65899a26dd6a16363194be4c2de2ff8` is complete: Core Build (all four unit-test shards, build, lint, types and guards), pg-real, pg-upgrade, tool-pg, CodeQL, security, compliance and CodeRabbit passed. There are no unresolved review threads or new actionable review comments.

Review follow-up: CodeRabbit's scope warning concerned a migration decision already present on the PR base. [CodeRabbit verified and acknowledged that the warning is not applicable](https://github.com/erp-mafia/accounted/pull/2602#issuecomment-5663494816); this PR has no migration diff. The [compliance follow-up](https://github.com/erp-mafia/accounted/pull/2602#issuecomment-5663507523) documents the input-precision audit and confirms the whole-ten tax base against IL 1 kap. 7 §. No further code changes were required.

Merged into main on 2026-09-14 as b620c28fbea4ef13e28acceed0119af32c9181a1. Issue #2597 is closed, and all review threads are resolved. The rounding source and test files exactly match the verified PR head. The [Vercel production deployment](https://vercel.com/jakobwennberg-gmailcoms-projects/erp-base/6ik5TB1WfYfVMzGz9BXr1d5Mwfpo) is READY and assigned to app.accounted.se. The earlier fork preview was not authorized; production deployed successfully from main.

### CODE REVIEW

File: lib/__tests__/whole-krona-drift.test.ts [hunk 1]
```diff
@@ -0,0 +1,63 @@
+import { describe, it, expect } from 'vitest'
+import { sruAmount } from '@/lib/reports/sru/format'
+import { calculateEgenavgifter } from '@/lib/bokslut/enskild-firma/egenavgifter-calculator'
+import { proposeAvsattning } from '@/lib/bokslut/reserves/periodiseringsfond-service'
```
[the remark this answers is not in the document]
Comment (valpet) [reply lib/__tests__/whole-krona-drift.test.ts:4]: Addressed: the INK2 regression now lives in `lib/reports/ink2/__tests__/ink2-declaration.test.ts`, where the trial-balance fixtures already exist. It plants three revenue accounts (1033.54 + 1203.03 + 259.43) that accumulate to 2495.99999999999955 and asserts `ink2r['7410']` is 2496. With the alias reverted to the old local `Math.floor` it fails with `expected 2495 to be 2496`.

Reviewing the rest of the class while I was in there turned up a fifth site the issue had missed: `bolagsskatt-calculator.ts:276` floors to whole *tens*, so the same drift costs 10 kr of taxable base rather than 1. That is fixed and covered here too.
